### PR TITLE
Fix memory leak in k8s state for deleted pods

### PR DESCRIPTION
**Pull request by scrosby**
## Changes proposed in this PR

- Fix a leak where we we would unconditionally write back to the k8s state when the state scanner visited a pod which had been deleted. This ends up leaking.

## Why are we making these changes?
- Avoids a memory leak.

### CODE REVIEW

File: scheduler/src/cook/kubernetes/controller.clj [hunk 1]
```diff
@@ -596,8 +596,12 @@
                     :synthesized-state (api/pod->synthesized-pod-state pod-name pod)
                     :sandbox-file-server-container-state (api/pod->sandbox-file-server-container-state pod)}
          old-state (get @k8s-actual-state-map pod-name)]
-     ; We always store the updated state, but only reprocess it if it is genuinely different.
-     (swap! k8s-actual-state-map assoc pod-name new-state)
+     ; We always store the new state (to capture any changes in the pod) if the pod's not nil. If we write here unconditionally,
+     ; we may leak a deleted pod back into this dictionary where it will never be recoverable.
+     ;
+     ; We only reprocess if something important has actually changed.
+     (when-not (and (nil? pod))
```
Comment (dposada) [inline scheduler/src/cook/kubernetes/controller.clj:603]:
```suggestion
     (when (some? pod)
```

File: scheduler/src/cook/kubernetes/controller.clj [hunk 2]
```diff
@@ -596,8 +596,12 @@
...
+       (swap! k8s-actual-state-map assoc pod-name new-state))
```
Comment (dposada) [inline scheduler/src/cook/kubernetes/controller.clj:604]: Can we write a unit test?

File: scheduler/src/cook/kubernetes/controller.clj (hunk 1 above)
Comment (scrosby) [reply scheduler/src/cook/kubernetes/controller.clj]: Thanks. Made.

File: scheduler/src/cook/kubernetes/controller.clj (hunk 2 above)
Comment (scrosby) [reply scheduler/src/cook/kubernetes/controller.clj]: Good suggestion. Done.

**Review by dposada — APPROVED** [review]

**Pull request merged**